pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -165,31 +165,38 @@ type resettableRESTMapper interface {
 // the course of detecting new resources.
 func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterface, period time.Duration, stopCh <-chan struct{}) {
 	oldResources := make(map[schema.GroupVersionResource]struct{})
+	syncInProgress := false
 	wait.Until(func() {
-		// Get the current resource list from discovery.
-		newResources := GetDeletableResources(discoveryClient)
+		syncAttemptDone := make(chan struct{})
+		defer close(syncAttemptDone)
+		newResources, resourcesUpdatedCh := listWatchDeletableResources(discoveryClient, period, syncAttemptDone)
 
 		// This can occur if there is an internal error in GetDeletableResources.
 		// If the gc attempts to sync with 0 resources it will block forever.
-		// TODO: Implement a more complete solution for the garbage collector hanging.
 		if len(newResources) == 0 {
 			glog.V(5).Infof("no resources reported by discovery, skipping garbage collector sync")
 			return
 		}
 
-		// Decide whether discovery has reported a change.
-		if reflect.DeepEqual(oldResources, newResources) {
-			glog.V(5).Infof("no resource updates from discovery, skipping garbage collector sync")
-			return
-		}
+		// If the sync is already in progress we don't need to check if discovery has reported a change,
+		// we will attempt to sync regardless.
+		if !syncInProgress {
+			// Decide whether discovery has reported a change.
+			if reflect.DeepEqual(oldResources, newResources) {
+				glog.V(5).Infof("no resource updates from discovery, skipping garbage collector sync")
+				return
+			}
 
-		// Something has changed, time to sync.
-		glog.V(2).Infof("syncing garbage collector with updated resources from discovery: %v", newResources)
+			// Something has changed, time to sync.
+			glog.V(2).Infof("syncing garbage collector with updated resources from discovery: %v", newResources)
 
-		// Ensure workers are paused to avoid processing events before informers
-		// have resynced.
-		gc.workerLock.Lock()
-		defer gc.workerLock.Unlock()
+			// Ensure workers are paused to avoid processing events before informers
+			// have resynced. We won't immediately defer an Unlock, however, because
+			// we cannot resume the workers unless the gc sync has completed successfully.
+			// If the sync doesn't complete, we will keep trying until it eventually succeeds.
+			gc.workerLock.Lock()
```

Comment:
This means we could lock gc workers for one period (30s) or longer. We probably need a for loop inside of `wait.Until(func, period, stopCh)` block to keep rediscovering and resyncing instead of waiting for the next round

Reply:
I don't think we should run discovery continuously... timing out waiting for cache sync and rediscovering at that point is probably sufficient (I was experimenting with a simpler loop retry in https://github.com/kubernetes/kubernetes/pull/64235/files?w=1)

Reply:
I think we are not running discovery continuously. 

If an API group gets removed after we start syncing cache, 1. `timing out waiting for cache sync and rediscovering at that point` (https://github.com/kubernetes/kubernetes/pull/64235) and 2. `periodically polling (listWatchDeletableResources) discovery after cache sync starts, and stopping cache sync if discovery result changes` (this implementation) work the same way if `1's timeout` = `2's period` (30s in both case). 

If it's network issue that causes cache sync to timeout, https://github.com/kubernetes/kubernetes/pull/64235 sends a new `gc.resyncMonitors(newResources)` request, and this implementation keeps waiting. 

From apiserver perspective, I don't see difference in discovery query.

My previous comment was about leaving `gc.workerLock` locked when `Sync()` is idle.

Reply:
ah, ok. once we call resyncMonitors with new resources, we should not unlock workers until we have successfully synced (the fact that we return now in that case and unpause workers is unsafe)